src/abstractdataframe/selection.jl

```diff
@@ -434,233 +576,24 @@ function select_transform!(@nospecialize(nc::Union{Base.Callable, Pair{<:Union{I
-    """
-
 """
     select!(df::DataFrame, args...; renamecols::Bool=true)
-    select!(args::Callable, df::DataFrame; renamecols::Bool=true)
-
-Mutate `df` in place to retain only columns specified by `args...` and return it.
-The result is guaranteed to have the same number of rows as `df`, except when no
-columns are selected (in which case the result has zero rows).
-
-$SELECT_ARG_RULES
-
-Note that including the same column several times in the data frame via renaming
-or transformations that return the same object without copying will create
-column aliases. An example of such a situation is
-`select!(df, :a, :a => :b, :a => identity => :c)`.
-
-# Examples
-```jldoctest
-julia> df = DataFrame(a=1:3, b=4:6)
-3×2 DataFrame
-│ Row │ a     │ b     │
-│     │ Int64 │ Int64 │
-├─────┼───────┼───────┤
-│ 1   │ 1     │ 4     │
-│ 2   │ 2     │ 5     │
-│ 3   │ 3     │ 6     │
-
-julia> select!(df, 2)
-3×1 DataFrame
-│ Row │ b     │
-│     │ Int64 │
-├─────┼───────┤
-│ 1   │ 4     │
-│ 2   │ 5     │
-│ 3   │ 6     │
-
-julia> df = DataFrame(a=1:3, b=4:6);
-
-julia> select!(df, :a => ByRow(sin) => :c, :b)
-3×2 DataFrame
-│ Row │ c        │ b     │
-│     │ Float64  │ Int64 │
-├─────┼──────────┼───────┤
-│ 1   │ 0.841471 │ 4     │
-│ 2   │ 0.909297 │ 5     │
-│ 3   │ 0.14112  │ 6     │
-
-julia> select!(df, :, [:c, :b] => (c,b) -> c .+ b .- sum(b)/length(b))
-3×3 DataFrame
-│ Row │ c        │ b     │ c_b_function │
-│     │ Float64  │ Int64 │ Float64      │
-├─────┼──────────┼───────┼──────────────┤
-│ 1   │ 0.841471 │ 4     │ -0.158529    │
-│ 2   │ 0.909297 │ 5     │ 0.909297     │
-│ 3   │ 0.14112  │ 6     │ 1.14112      │
-
-julia> df = DataFrame(a=1:3, b=4:6);
-
-julia> select!(df, names(df) .=> [minimum maximum]);
-
-julia> df
-3×4 DataFrame
-│ Row │ a_minimum │ b_minimum │ a_maximum │ b_maximum │
-│     │ Int64     │ Int64     │ Int64     │ Int64     │
-├─────┼───────────┼───────────┼───────────┼───────────┤
-│ 1   │ 1         │ 4         │ 3         │ 6         │
-│ 2   │ 1         │ 4         │ 3         │ 6         │
-│ 3   │ 1         │ 4         │ 3         │ 6         │
-
-julia> df = DataFrame(a=1:3, b=4:6);
-
-julia> using Statistics
-
-julia> select!(df, AsTable(:) => ByRow(mean), renamecols=false)
-3×1 DataFrame
-│ Row │ a_b     │
-│     │ Float64 │
-├─────┼─────────┤
-│ 1   │ 2.5     │
-│ 2   │ 3.5     │
-│ 3   │ 4.5     │
+    select!(args::Base.Callable, df::DataFrame; renamecols::Bool=true)
+    select!(gd::GroupedDataFrame{DataFrame}, args...; ungroup::Bool=true, renamecols::Bool=true)
+    select!(f::Base.Callable, gd::GroupedDataFrame; ungroup::Bool=true, renamecols::Bool=true)
 
-julia> df = DataFrame(a=1:3, b=4:6);
+Mutate `df` or `gd` in place to retain only columns specified by `args...` and
+return it. The result is guaranteed to have the same number of rows as `df` or
+parent of `gd`, except when no columns are selected (in which case the result
+has zero rows).
```

Comment:
You mean to disallow `select(df)`?, `select(groupby(df, cols), keepkeys=false)` and `select(groupby(df, []))`?

Then `select(groupby(df, cols))` would pass or fail depending on the contents of `cols`.

I am not sure what is best - internally we treat data frame with 0 columns as having an undefined number of rows. We return `0` only because we have to return something when calling `nrow`.

Reply:
Yeah. Calling `select` without any transformation doesn't sound terribly useful. Though I guess it's good to have for full generality. Indeed, otherwise you wouldn't be able to call `select` on a zero-column data frame...

Reply:
I will leave it as is.